runtime/lua/vim/lsp/inlay_hint.lua

```diff
@@ -0,0 +1,253 @@
+local util = require('vim.lsp.util')
+local log = require('vim.lsp.log')
+local api = vim.api
+local M = {}
+
+-- [[
+-- TODO:
+-- [ ] only request for visible range
+-- ]]
+--
+---@class lsp.inlay_hint.bufstate
+---@field version integer
+---@field client_hint table<integer, table<integer, lsp.InlayHint[]>> client_id -> (lnum -> hints)
+
+---@type table<integer, lsp.inlay_hint.bufstate>
+local hint_cache_by_buf = setmetatable({}, {
+  __index = function(t, b)
+    local key = b > 0 and b or api.nvim_get_current_buf()
+    return rawget(t, key)
+  end,
+})
+
+--- |lsp-handler| for the method `textDocument/inlayHint`
+--- Store hints for a specific buffer and client
+--- Resolves unresolved hints
+---
+function M.on_inlayhint(err, result, ctx, _)
+  if err then
+    local _ = log.error() and log.error('inlayhint', err)
+    return
+  end
+  local bufnr = ctx.bufnr
+  local client_id = ctx.client_id
+  local version = ctx.version
+  if not result then
+    return
+  end
+  local bufstate = hint_cache_by_buf[bufnr]
+  if not bufstate then
+    bufstate = {
+      client_hint = vim.defaulttable(),
+      version = version,
+    }
+    hint_cache_by_buf[bufnr] = bufstate
+    api.nvim_buf_attach(bufnr, false, {
+      on_detach = function(_, b)
+        hint_cache_by_buf[b] = nil
+      end,
+      on_reload = function(_, b)
+        hint_cache_by_buf[b] = nil
+      end,
+    })
+  end
+  local hints_by_client = bufstate.client_hint
+  local hints_by_lnum = hints_by_client[client_id]
+  local client = vim.lsp.get_client_by_id(client_id)
+
+  -- To reduce flicker, preserve old resolved hints until new ones are resolved.
+
+  local new_hints_by_lnum = vim.defaulttable()
+  local num_unprocessed = #result
+  if num_unprocessed == 0 then
+    hints_by_client[client_id] = {}
+    bufstate.version = version
+    return
+  end
+
+  for _, hint in ipairs(result) do
+    local lnum = hint.position.line
+    table.insert(new_hints_by_lnum[lnum], hint)
+  end
+
+  for lnum, _ in pairs(hints_by_lnum) do
+    if not next(new_hints_by_lnum[lnum]) then
+      hints_by_lnum[lnum] = nil
+    end
+  end
+
+  local countdown = function()
+    num_unprocessed = num_unprocessed - 1
+    if num_unprocessed == 0 then
+      bufstate.version = version
```

Comment:
This is where the cache's version is set to the requested version. But it only runs once all the unresolved hints have been fully resolved. Before the decoration provider draws anything, this has to be set.

My comment was that the cache's version should be updated no matter what by the initial response handler, while you fire off the requests to resolve anything that's needs it.

This makes sure the decoration provider can immediately use this result, even before additional info from the hints are still pending.

As for what to resolve and how to track it, maybe a future PR? I'm not sure how commands should be used, and tooltip seems like it'd need tighter integration with the built in hover function?